README.md

```diff
@@ -1,10 +1,300 @@
+* [`Value`][Value] - For services that always return a static value.
+* [`Alias`][Alias] - An alias for another service, with defaulting capabilities for when the original does not exist.
+* [`GlobalVar`][GlobalVar] - For services that return global variables.
+
+# Mutation
+
+The [`withDependencies()`][withDeps] method allows all service instances to be copied with different dependencies, while leaving the origianl instances unaffected.
+
+```php
+$service = new Factory(['database'], function ($database) {
+  // ...
+});
+
+$service2 = $service->withDependencies(['db']);
+```
+
+This makes it possible to modify service dependencies at run-time, or even during a build process, which can be especially useful when dealing with 3rd party service providers that need to be rewired.
+
+# Multi-boxing
+
+One of the benefits of being able to derive new services with different dependencies is the ability to use the same provider multiple times. Let's look at an example.
+
+Consider a service provider for a logger that writes to a file.
+
+```php
+class LoggerProvider implements ServiceProviderInterface
+{
+  public function getFactories()
+  {
+    return [
+      'logger' => new Constructor(FileLogger::class, ['file_path']),
+      'file_path' => new Value(sys_get_tmp_dir() . '/log.txt')
+    ];
+  }
+  
+  public function getExtensions ()
+  {
+    return [];
+  }
+}
+```
+
+Our application needs to keep 2 different log files: one for errors and one for debugging.
+
+Simply using the above service provider twice won't work; we'd be re-declaring the `logger` and `file_path` services.
+
+Prefixing the factories _would_ allow us to have two instances of the service provider, but that would break the dependencies. If we prefixed the factories from one logger such that they become `debug_logger` and `debug_file_path`, the `debug_logger` factory would still be depending on `file_path`, which would no longer exist after prefixing.
+
+This where mutation of dependencies comes in. We can write a `PrefixingProvider` decorator that not only prefixes all services in a provider, but also prefixes any dependencies.
+
+_(The below class is incomplete for the sake of brevity. Assume that a constructor exists and that it initializes its `$prefix` and `$provider` properties)_.
+
+```php
+class PrefixingProvider implements ServiceProviderInterface {
+  public function getFactories() {
+    $factories = [];
+
+    foreach ($this->provider->getFactories() as $key => $factory) {
+      $deps = $factory->getDependencies();
+      $newDeps = array_map(fn($dep) => $this->prefix . $dep, $deps);
+
+      $factories[$this->prefix . $key] = $factory->withDependencies($newDeps);
+    }
+    
+    return $factories;
+  }
+}
+```
+
+We can now create two different versions of the same service provider:
+
+```php
+$debugLogProvider = new PrefixingProvider('debug_', new LoggerProvider);
+$errorLogProvider = new PrefixingProvider('error_', new LoggerProvider);
```

Comment:
Of course, this would mean prefixing of _all_ dependencies of _all_ factories. In practice, when creating a modular system where a module declares a lot of different configuration, this is not a good idea: in this example, services that do not depend on logging should remain, and this is not possible to reliably do by using the service key alone. Of course, it's still a useful capability, there's nothing wrong with that. However, I don't think that it _alone_ is enough to allow multi-boxing.

Reply:
The current problem that makes multi-boxing impossible is the fact that modules depend on **real** keys, as opposed to **virtual** keys. Allow me to explain:

Lets say module A has 1 or more services that depend on some external `some_service`, and module B graciously provides it. This alone is fine - inter-service dependencies is what makes DIC-powered apps work. But it also means that module B **must** make sure that `some_service` will absolutely remain keyed as `some_service`, otherwise it doesn't fulfill the requirement for module A.

And if the service keys must remain fixed, to avoid breaking dependents, module B can only be used once. Using module B twice, as instances X and Y, means having the exact same services from instance Y overriding all of the services from instance X (or vice versa depending on which is loaded first) and introduces the problem of having extensions being applied twice (since those are not overridden but concatenated).

The solution, in my opinion, is to have modules depend on **virtual** keys which are generated using app-level prefixing, rather than hard-requiring **real** keys. This also means that modules should not be aware of any external services or modules.

This can be seen in the logger example, where a service in a module does not depend on any `logger`, but a purpose-specific one using a virtual key such as `debug_logger`. This integration may be done automatically simply by virtue of having modules satisfy each other's dependencies, by the app explicitly aliasing some services, or heck even by re-using a module with the right prefix.

As a result, multi-boxing _is most definitely_ achieved, and control is once more handed back to the application.

Reply:
I see this a lot clearer now, thank you!